### PR TITLE
Generic Action\Action Button System

### PR DESCRIPTION
**Pull request by AnturK**
#### Adds Generic Button and Action System

/datum/action - instantiated once per action allowed then assigned to mobs using Grant()
/obj/screen/movable/action_button - generic button with modifiable icon/background, now movable if you're into that.

Currently supports:
 Item Actions
 Spells
 Works for all mob/living
##### TODO:

Actual Sprites / not eyesore colors
Helpers for adding generic actions

dd
im
space station 13 2015-04-05 180539
space station 13 2015-04-05 180626
space station 13 2015-04-05 180837
space station 13 2015-04-10 223423
space station 13 2015-04-10 222753

### CODE REVIEW

**Comment by optimumtact** [comment]
Nice art :smile: 

Good work though

**Comment by optimumtact** [comment]
This might be a little out of scope, but how easy would it be to make an action button that spawns another group of action buttons? So you can hide/show buttons

**Comment by AnturK** [comment]
That's how i wanted to make current ones work but i got lazy, probably doable will try tommorow. Grouped buttons is nice idea.

**Comment by xxalpha** [comment]
Can you have more than one action button per item? Can you have action buttons for items inside items inside mob.contents?

This is important, please respond.

**Comment by AnturK** [comment]
Yes but you need to add the additional ones manually and checks are flag based so i'd just need to add flag for that

**Comment by RemieRichards** [comment]
0/10 not as cool as VG's unless it uses my movable UI system.

**Comment by AnturK** [comment]
Visual themes were next thing on my list but that

**Comment by MrPerson** [comment]
Extensions can be done later.

Coderart is best art.

File: code/_onclick/hud/action.dm [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	T.update_action_buttons()
+	owner = null
+	return
+
+/datum/action/proc/Trigger()
+	world << "[src] Trigger"
+	if(!Checks())
+		world << "[src] Checks Failed"
+		return
+	world << "[src] Passed Checks"
```
Comment (MrPerson) [inline code/_onclick/hud/action.dm]: You left some world << "stuff" in there. NBD just get rid of them

File: code/_onclick/hud/action.dm [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+#define AB_CHECK_LYING 4
+#define AB_CHECK_ALIVE 8
+#define AB_CHECK_INSIDE 16
+
+
+/datum/action
+	var/name = "Generic Action"
+	var/action_type = AB_ITEM
+	var/procname = null
+	var/target = null
```
Comment (MrPerson) [inline code/_onclick/hud/action.dm]: I feel like target should var/obj/target or at least var/atom/movable/target.

File: code/_onclick/hud/action.dm (hunk 1 above)
Comment (optimumtact) [inline code/_onclick/hud/action.dm]: [WIP]

File: code/_onclick/hud/action.dm [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+#define AB_MAX_COLUMNS 10
+
+/datum/hud/proc/ButtonNumberToScreenCoords(var/number) // TODO : Make this zero-indexed for readabilty
+	var/row = round(number/(AB_MAX_COLUMNS+1))
+	var/col = ((number - 1)%(AB_MAX_COLUMNS)) + 1
+	var/coord_col = col==1?"":"+[col-1]"
+	var/coord_col_offset = 4+2*col
+	var/coord_row = "[-1 - row]"
+	var/coord_row_offset = 26
+	return "WEST[coord_col]:[coord_col_offset],NORTH[coord_row]:[coord_row_offset]"
```
Comment (MrPerson) [inline code/_onclick/hud/action.dm]: You can just do WEST+0:33 instead of WEST+1:1

**Comment by MrPerson** [comment]
If I were doing this, I'd wind up making it way more complicated than this for absolutely no reason. Good stuff :+1:

**Comment by Fox-McCloud** [comment]
:+1:

File: code/_onclick/hud/action.dm (hunk 3 above)
Comment (AnturK) [inline code/_onclick/hud/action.dm]: Did that then found out screen loc and transform are not actually exclusive. Oh well

File: code/_onclick/hud/action.dm (hunk 2 above)
Comment (RemieRichards) [inline code/_onclick/hud/action.dm]: :+1: for ```var/atom/movable/target```

Comment (AnturK) [inline code/_onclick/hud/action.dm]: Oh forgot about this, will update in a moment.

**Comment by RemieRichards** [comment]
The movable elements would probably be nicer as /movable/snap, up to you.

**Comment by AnturK** [comment]
Blarg. Started working on turning alien abilities into proc_holders but then realized that might be bit out of scope for one PR so i'll leave that for later.

**Comment by AnturK** [comment]
image
image
image
image

Current look for the backgrounds and show button

**Comment by AnturK** [comment]
This is merge ready, not counting better button icons. I'd do the changeling abilities buttons but i'm starting to question my assumption that ugly buttons are better than none.

**Pull request merged**